### PR TITLE
Fix bugs in ModuleFileSetBuilder and Workspace

### PR DESCRIPTION
**Pull request by bufdev**
See code comments. This is also starting to undo the mess in the workspace code, this is a long story.

We don't have super high confidence in this PR, but we think it puts us in a better state than we were in. It may cause bugs, although I have extensively QA'ed this.

### CODE REVIEW

**Review by doriable — APPROVED** [review]

File: private/buf/bufwork/workspace_builder.go [hunk 1]
```diff
@@ -64,9 +60,6 @@ func (w *workspaceBuilder) BuildWorkspace(
 	allModules := make([]bufmodule.Module, 0, len(workspaceConfig.Directories))
 	for _, directory := range workspaceConfig.Directories {
 		if cachedModule, ok := w.moduleCache[directory]; ok {
-			if directory == targetSubDirPath {
```
Comment (bufdev) [inline private/buf/bufwork/workspace_builder.go:67]: For posterity: workspaces were being created that specifically *excluded* the "target" module - this caused other bugs down the line. I have no idea why this worked, but a workspace should include all the modules within it.

File: private/buf/bufwork/workspace_builder.go [hunk 2]
```diff
@@ -28,16 +28,12 @@ import (
 )
 
 type workspaceBuilder struct {
-	moduleBucketBuilder bufmodulebuild.ModuleBucketBuilder
-	moduleCache         map[string]*cachedModule
+	moduleCache map[string]*cachedModule
 }
 
-func newWorkspaceBuilder(
-	moduleBucketBuilder bufmodulebuild.ModuleBucketBuilder,
```
Comment (bufdev) [inline private/buf/bufwork/workspace_builder.go:36]: For posterity: this `ModuleBucketBuilder` was never used.

File: private/buf/bufwork/bufwork.go [hunk 3]
```diff
@@ -98,7 +98,9 @@ var (
 // WorkspaceBuilder builds workspaces. A single WorkspaceBuilder should NOT be persisted
 // acorss calls because the WorkspaceBuilder caches the modules used in each workspace.
 type WorkspaceBuilder interface {
-	// BuildWorkspace builds a bufmodule.Workspace for the given targetSubDirPath.
+	// BuildWorkspace builds a bufmodule.Workspace.
+	//
+	// The given targetSubDirPath is the only path that will have the configOverride applied to it.
```
Comment (bufdev) [inline private/buf/bufwork/bufwork.go:103]: We likely want to eventually eliminate `targetSubDirPath` from workspace construction.

File: private/bufpkg/bufmodule/workspace.go [hunk 4]
```diff
@@ -23,13 +28,30 @@ type workspace struct {
 }
 
 func newWorkspace(
+	ctx context.Context,
 	namedModules map[string]Module,
 	allModules []Module,
-) *workspace {
+) (*workspace, error) {
+	pathToExternalPaths := make(map[string][]string)
```
Comment (bufdev) [inline private/bufpkg/bufmodule/workspace.go:35]: This enforces the property that workspaces must have unique files

**Review by doriable — APPROVED** [review]

**Pull request merged**